app/models/concerns/master_listable.rb

```diff
@@ -0,0 +1,181 @@
+    before_validation :set_master_list,   unless: :is_master_list?
+    before_save :abort_if_master_changed
+    before_save :remove_master_list_id,   if: :is_master_list?
+    before_save :set_title_to_master,     if: :is_master_list?
+    before_destroy :abort_if_master,      if: :has_child_lists?
+    after_destroy :destroy_master_list,   unless: -> { is_master_list? || master_has_other_children? }
+  end
+
+  def add_item_from_child_list(item)
+    raise MasterListError, 'add_item_from_child_list method only available on master lists' unless is_master_list?
+
+    list_items.combine_or_create!(public_list_item_attrs(item).merge('list_id' => id))
+  end
+
+  def remove_item_from_child_list(attrs)
+    raise MasterListError, 'remove_item_from_child_list method only available on master lists' unless is_master_list?
+
+    existing_item = list_items.find_by(description: attrs['description'])
+
+    if existing_item.nil? || existing_item.quantity < attrs['quantity']
+      raise MasterListError, 'item passed to remove_item_from_child_list method is not represented on the master list'
+    end
+
+    if existing_item.quantity == attrs['quantity']
+      existing_item.destroy!
+    else
+      existing_item.quantity -= attrs['quantity']
+      existing_item.notes = extract_notes(attrs['notes'], existing_item.notes)
+      existing_item.save!
+    end
+
+    existing_item&.persisted? ? existing_item : nil
+  end
+
+  def update_item_from_child_list(description, delta_quantity, old_notes, new_notes)
+    raise MasterListError, 'update_item_from_child_list method only available on master lists' unless is_master_list?
+
+    existing_item = list_items.find_by(description: description)
+
+    if existing_item.nil? || delta_quantity < (-existing_item.quantity)
+      raise MasterListError, 'invalid data to update master list item'
+    end
+
+    existing_item.quantity += delta_quantity
+    existing_item.notes = if old_notes.nil? && new_notes.present?
+                            [existing_item.notes.to_s, new_notes.to_s].join(' -- ')
+                          else
+                            existing_item.notes&.sub(/#{old_notes}/, new_notes.to_s).presence || new_notes
+                          end
+
+    existing_item.save!
+    existing_item
+  end
+
+  def extract_notes(notes, existing)
+    return existing unless notes && existing.to_s =~ /#{notes}/
+
+    existing.sub(notes, '')
+  end
+
+  def ensure_master_list_is_master
+    if master_list&.master != true
+      errors.add(:master_list, 'must be a master list')
+    end
+  end
+
+  def set_master_list
+    self.master_list ||= self.class.find_or_create_by!(user: user, master: true)
+  end
+
+  def create_master_list
+    self.class.find_or_create_by!(user: user, master: true)
+  end
+
+  def set_title_to_master
+    self.title = 'Master'
+  end
+
+  def abort_if_master_changed
+    throw :abort if master_changed? && !new_record?
```

Comment:
It'll also be marked as changed when the record is new so have to check that it isn't the initial create.